The roots of empty transaction hash subtrees are now uniformly committed to with empty hashes. (#505)
zcash/zips: zip-0244.rst

T.2: transparent_digest
```````````````````````

> In the case that transparent inputs or outputs or present, the transparent digest

Reviewer: ```suggestion
In the case that transparent inputs or outputs are present, the transparent digest
```

  "ZTxIdPrevoutHash"

> In the case that the transaction has no transparent inputs, ``prevouts_digest`` is ::

Reviewer 2: ```suggestion
In the case that the transaction has transparent outputs but no transparent inputs,
``prevouts_digest`` is ::
```

  "ZTxIdSequencHash"

> In the case that the transaction has no transparent inputs, ``sequence_digest`` is ::

Reviewer 2: ```suggestion
In the case that the transaction has transparent outputs but no transparent inputs,
``sequence_digest`` is ::
```

  BLAKE2b-256("ZTxIdSequencHash", [])

T.2c: outputs_digest
''''''''''''''''''''
A BLAKE2b-256 hash of the concatenated field encodings of all transparent
output values of the transaction. The field encoding of such an output consists
of the encoded output ``amount`` (8-byte little endian) followed by
the ``scriptPubKey`` byte array (serialized as Bitcoin script).

The personalization field of this hash is set to::

  "ZTxIdOutputsHash"

> In the case that the transaction has no transparent outputs, ``outputs_digest`` is ::

Reviewer 2: ```suggestion
In the case that the transaction has transparent inputs but no transparent outputs,
``outputs_digest`` is ::
```

  "ZTxIdSSpendsHash"

> In the case that no Sapling spends are present, ``sapling_spends_digest`` is ::

Reviewer 2: ```suggestion
In the case that the transaction has Sapling outputs but no Sapling spends,
``sapling_spends_digest`` is ::
```

  "ZTxIdSOutputHash"

> In the case that no Sapling outputs are present, ``sapling_outputs_digest`` is ::

Reviewer 2: ```suggestion
In the case that the transaction has Sapling spends but no Sapling outputs,
``sapling_outputs_digest`` is ::
```

The personalization field of this hash is set to::

  "ZTxIdOrcActMHash"

T.4c: orchard_actions_noncompact_digest
"""""""""""""""""""""""""""""""""""""""

A BLAKE2b-256 hash of the remaining subset of Orchard Action information **not** intended
for inclusion in an updated version of the the ZIP 307 [#zip-0307]_ ``CompactBlock``
format, for all Orchard Actions belonging to the transaction. For each Action,
the following elements are included in the hash::

   T.4c.i  : cv                    (field encoding bytes)
   T.4c.ii : rk                    (field encoding bytes)
   T.4c.iii: encCiphertext[564..]  (post-memo suffix of field encoding)
   T.4c.iv : outCiphertext         (field encoding bytes)

The personalization field of this hash is set to::

  "ZTxIdOrcActNHash"

Signature Digest
================

A new per-input transaction digest algorithm is defined that constructs a hash that may be
signed by a transaction creator to commit to the effects of the transaction. A signature
digest is produced for each transparent input, each Sapling input, and each Orchard
action. For transparent inputs, this follows closely the algorithms from ZIP 143 [#zip-0143]_

> and ZIP 243 [#zip-0243]_. For other input types, the txid may be signed directly.

Reviewer 3: ```suggestion
and ZIP 243 [#zip-0243]_. For other input types, this algorithm has the exact same
output as the transaction digest algorithm, thus the txid may be signed directly.
```

The personalization field of this hash is set to::

  "ZTxIdOrcActMHash"

T.4c: orchard_actions_noncompact_digest
"""""""""""""""""""""""""""""""""""""""

A BLAKE2b-256 hash of the remaining subset of Orchard Action information **not** intended
for inclusion in an updated version of the the ZIP 307 [#zip-0307]_ ``CompactBlock``
format, for all Orchard Actions belonging to the transaction. For each Action,
the following elements are included in the hash::

   T.4c.i  : cv                    (field encoding bytes)
   T.4c.ii : rk                    (field encoding bytes)
   T.4c.iii: encCiphertext[564..]  (post-memo suffix of field encoding)
   T.4c.iv : outCiphertext         (field encoding bytes)

The personalization field of this hash is set to::

  "ZTxIdOrcActNHash"

Signature Digest
================

A new per-input transaction digest algorithm is defined that constructs a hash that may be
signed by a transaction creator to commit to the effects of the transaction. A signature
digest is produced for each transparent input, each Sapling input, and each Orchard
action. For transparent inputs, this follows closely the algorithms from ZIP 143 [#zip-0143]_

> and ZIP 243 [#zip-0243]_. For other input types, this algorithm has the exact same
> output as the transaction digest algorithm, thus the txid may be signed directly.

Reviewer 2: ```suggestion
and ZIP 243 [#zip-0243]_. For shielded inputs, this algorithm has the exact same output
as the transaction digest algorithm, thus the txid may be signed directly.
```

otherwise::

   BLAKE2b-256(``ZTxIdSequencHash``, [])

S.2c: outputs_sig_digest
''''''''''''''''''''''''
This is a BLAKE2b-256 hash initialized with the personalization field value

> ``ZTxIdOutputsHash``. 

Reviewer 2: Revert this whitespace change.

   the hash is over the transaction serialized form of the transparent output at that
   index

If the sighash type is ``SIGHASH_SINGLE`` and the signature is being computed for
a shielded input, or if the sighash type is ``SIGHASH_NONE``::
otherwise::

   the hash is immediately finalized, without being updated with any
   additional data
   BLAKE2b-256(``ZTxIdOutputsHash``, [])

S.2d: txin_sig_digest
'''''''''''''''''''''
This is a BLAKE2b-256 hash initialized with the personalization field value
``Zcash___TxInHash`` (3 underscores).

If the signature hash is being computed for a transparent input, the hash
is updated with the following properties of that input::
This is a BLAKE2b-256 hash of the following properties of the transparent input being
signed, initialized with the personalization field value ``Zcash___TxInHash`` (3
underscores).

   S.2d.i:   prevout     (field encoding)
   S.2d.ii:  script_code (field encoding)
   S.2d.iii: value       (8-byte signed little-endian)
   S.2d.iv:  nSequence   (4-byte unsigned little-endian)

otherwise::

   the hash is immediately finalized, without being updated with any
   additional data

> 

Reviewer: Should there be an "otherwise: empty hash" here as well?

Author: It is not necessary, because if the input being signed is not a transparent input, S.2 specifies that the hash be identical to T.2

Reviewer 3: Yep, so this sub-function will never be called except when there is a transparent input that requires digesting.

the transparent input at a particular index, and a transparent output appears in
the transaction at that index::

   the hash is over the transaction serialized form of the transparent output at that
   index

otherwise::

   BLAKE2b-256(``ZTxIdOutputsHash``, [])

S.2d: txin_sig_digest
'''''''''''''''''''''
This is a BLAKE2b-256 hash of the following properties of the transparent input being
signed, initialized with the personalization field value ``Zcash___TxInHash`` (3
underscores).

   S.2d.i:   prevout     (field encoding)
   S.2d.ii:  script_code (field encoding)

>    S.2d.iii: value       (8-byte signed little-endian)

Reviewer 3: Hmm, I just spotted this. The value here should be unsigned (as we cannot have coins with negative values).

In practice it doesn't make much difference, as the upper range of a coin value would get coerced to the same encoding as signed integers, but is completely out of range for MAX_MONEY so won't ever be encountered. And I presume this snuck in because the `CAmount` type in `zcashd` is an `int64_t` (because it is dual-purposed both for non-negative and negative-allowed uses).

Reviewer: Zebra also uses `i64` internally to convert to bytes:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/amount.rs#L49

But our `NonNegative` constraint ensures that the value is positive and below `MAX_MONEY`:
https://github.com/ZcashFoundation/zebra/blob/main/zebra-chain/src/amount.rs#L336

Reviewer 2: If it's required to be nonnegative then it should be described as unsigned.

Reviewer 2: @nuttycom and @dconnolly and I reviewed this and think it's fine as-is.
